presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -393,6 +398,25 @@ public void addResetSessionProperties(String name)
         resetSessionProperties.add(requireNonNull(name, "name is null"));
     }
 
+    public Map<String, String> getAddedPreparedStatements()
+    {
+        return addedPreparedStatements;
+    }
+
+    public void addPreparedStatement(String key, String value)
+    {
+        requireNonNull(key, "key is null");
+        requireNonNull(value, "value is null");
+
+        if (session.getPreparedStatements().containsKey(key)) {
```

Comment:
putIfAbsent goes to the addedPreparedStatements, which are the things to be updated as a result of this transaction.  But first you need to check that the session doesn't already have a prepared statement with that name, or you'll end up overwriting an existing statement, which is not allowed.